### ISSUE TITLE
3x Performance Regression of Transposed MatMul over Regular MatMul

### ISSUE THREAD

**Issue by dmholtz** [torch-spyre/torch-spyre#3512]
## Summary

On the `spyre` device, the transposed matrix multiplication (`x @ A.T`) is **~3.5x slower** than regular matrix multiplication (`x @ A`) for identical shapes. This directly impacts `nn.Linear`, which implements `y = x A^T + b`, causing a critical performance regression for all standard neural network inference workloads.

---

## Environment

- **Device:** `spyre`
- **Platfrom:** `s390x`
- **dtype:** `torch.float16`

---

## Reproducer

```python
import torch
import torch.nn.functional as F
from torch.profiler import record_function

# shapes
B = 1
M = 4096
N = 51200

device = "spyre"
print(f"Running benchmark on device={device}")

dtype = torch.float16

num_warmup = 2
num_steps = 10

X   = torch.rand((B, M), dtype=dtype, device=device)
A   = torch.rand((M, N), dtype=dtype, device=device)
A_T = torch.rand((N, M), dtype=dtype, device=device)

@torch.compile
def op_matmul(x, a):
    return x @ a

@torch.compile
def op_transposed_matmul(x, a):
    return x @ a.T

with torch.profiler.profile(
    activities=[
        torch.profiler.ProfilerActivity.CPU,
        torch.profiler.ProfilerActivity.PrivateUse1,
    ],
    schedule=torch.profiler.schedule(
        wait=0, warmup=num_warmup, active=num_steps - num_warmup, repeat=1
    ),
) as p:
    for i in range(num_steps):
        with record_function("regular_matmul"):
            y_reg = op_matmul(X, A)

        with record_function("transposed_matmul"):
            y_transpose = op_transposed_matmul(X, A_T)

        p.step()

print(p.key_averages().table(sort_by="cpu_time_total", row_limit=-1))
```

> **Shapes are inspired from the Granite 3.3 8B model**, reflecting realistic production sizes.

---

## Profiler Output

```
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                                   Name    Self CPU %      Self CPU   CPU total %     CPU total  CPU time avg    # of Calls
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
                                          ProfilerStep*         0.10%     121.666us       100.00%     117.637ms      14.705ms             8
                                      transposed_matmul         0.13%     151.227us        77.29%      90.919ms      11.365ms             8
                             Torch-Compiled Region: 1/0         0.15%     172.965us        77.10%      90.695ms      11.337ms             8
## Call CompiledFxGraph fo2socux4fnagrd62hb4jlbqn2pv...         0.16%     183.940us        76.92%      90.490ms      11.311ms             8
                 launch_jobplan:sdsc_fused_mm_permute_0        76.77%      90.306ms        76.77%      90.306ms      11.288ms             8
                                         regular_matmul         0.13%     156.360us        22.61%      26.596ms       3.324ms             8
                             Torch-Compiled Region: 0/0         0.16%     185.733us        22.41%      26.358ms       3.295ms             8
## Call CompiledFxGraph fuceaupkeoqx2nm6hjfoisco72yi...         0.11%     128.023us        22.22%      26.138ms       3.267ms             8
                         launch_jobplan:sdsc_fused_mm_0        22.11%      26.010ms        22.11%      26.010ms       3.251ms             8
                               TorchDynamo Cache Lookup         0.13%     153.861us         0.13%     153.861us       9.616us            16
                 AOTDispatcher Runtime Wrapper Prologue         0.03%      36.571us         0.03%      36.571us       2.286us            16
                                      Pregraph bytecode         0.03%      29.932us         0.03%      29.932us       1.871us            16
-------------------------------------------------------  ------------  ------------  ------------  ------------  ------------  ------------
```

### Key Numbers

| Operation | CPU time avg | Kernel dispatched |
|---|---|---|
| `regular_matmul` (`x @ A`) | **3.251 ms** | `sdsc_fused_mm_0` |
| `transposed_matmul` (`x @ A.T`) | **11.288 ms** | `sdsc_fused_mm_permute_0` |
| **Regression factor** | **~3.47×** | — |

The profiler makes the root cause visible: the transposed path dispatches a different kernel, `sdsc_fused_mm_permute_0`, instead of `sdsc_fused_mm_0`. The extra `permute` step carries a significant and disproportionate overhead.

---

### 1. Parity with other backends

CUDA and Metal handle `x @ A.T` at native matmul speed through cuBLAS/MPS transpose flags — no data movement required. The spyre backend should reach the same parity.

### 2. `nn.Linear` is affected by default

`torch.nn.Linear` implements:

$$y = x A^T + b$$

When weights are loaded from a checkpoint, `weight` is stored with shape `(out_features, in_features)`, so every forward pass of every `nn.Linear` layer is a transposed matmul. This means the regression is not an edge case — **it affects every standard model from the PyTorch ecosystem**.

### 3. Impact on realistic models

The benchmark shapes (`M=4096, N=51200`) are taken directly from **Granite 3.3 8B** projection layers. A ~3.5× slowdown on each `nn.Linear` call compounds across hundreds of layers per forward pass, making end-to-end inference latency unacceptable.

---

## Workaround (Not Scalable)

This workaround was implemented in **spyre-inference** for Granite 3.3 8B. It pre-transposes the weight matrix once during model loading, so the cost is paid only once and does not affect inference latency:

```python
# Instead of (called at every forward pass):
y = F.linear(x, weight)          # weight shape: (N, M)  →  dispatches sdsc_fused_mm_permute

# Transpose once at model load time:
weight_T = weight.T.contiguous() # weight_T shape: (M, N)  — done once, not per inference step

# Then at every forward pass (fast path):
y = x @ weight_T                 # dispatches sdsc_fused_mm
```

This reproduces the ~3× speedup for the affected linear layers seen in the dummy benchmark. However, it requires **per-model manual surgery** and does not scale to the broader HuggingFace model zoo.

---

## Expected Fix

The spyre backend should handle `x @ A.T` as efficiently as `x @ A`, transparently and without requiring changes to user models. This would fix [`nn.Linear`](https://pytorch.org/docs/stable/generated/torch.nn.Linear.html) and [`F.linear`](https://pytorch.org/docs/stable/generated/torch.nn.functional.linear.html) across the entire PyTorch model ecosystem without any downstream workarounds.

---

## Additional Context

- Confirmed on shapes derived from **Granite 3.3 8B** (`M=4096, N=51200`).
- Regression is observable after `torch.compile` warmup (profiler excludes compilation overhead).
- No regression on the **shape** or **dtype** side — both ops use `float16`, same batch and sequence dimensions.

**Comment by mudhakar** [comment]
@dmholtz: A load-time fix already exists and is on by default: `model_utils.py` (#1339) sets `dim_order=[1,0]` on the weight's *device* layout, so the DMA stickifies along `out_features` on transfer. Shape stays `(out, in)`, `F.linear` unmodified, the `transpose` becomes metadata-only — no CPU transpose, no extra copy. `_monkey_patch.py:230-239` installs it globally.

So this is a coverage gap in the load path, not a missing backend capability. Two independent blockers:

1. **Policy too narrow** — `_transfer_module` gates on `isinstance(module, nn.Linear)` (`model_utils.py:178,185`); vLLM's parallel linear classes don't derive from it, so QKV/MLP weights get the default layout. (Not TP — spyre-inference is TP=1-only.) hf-adapters uses the broader rule at `hf_common.py:1199`: all 2-D weights except `nn.Embedding.weight`.
2. **`_apply` overrides bypass it** — `_transfer_module` returns early for those with a hardcoded default-layout lambda (`model_utils.py:166-172`). `SpyreParallelLMHead` overrides `_apply` (`parallel_lm_head.py:118-121`), so the LM head stays slow regardless of (1).

@JRosenkranz — could you take 1 and 2?

@tdoublep @bringlein — a question on the vLLM side: even with 1 and 2 fixed, spyre-inference inherits the layout implicitly through `nn.Module.to` internals, and `_apply` overrides like the LM head's are what break that. `process_weights_after_loading` is a no-op today (`custom_ops/linear.py:70`). Is that the better place to own weight layout for the vLLM path, or is there a reason to keep it implicit?

**Comment by dmholtz** [comment]
@mudhakar Thank you for the clarifications. It looks like I ran into an edge case rather than a fundamental flaw.

Regarding (1), I indeed tested my workaround on SpyreParallelLMHead, confirming the coverage gap. I'll take a look at the other layers tomorrow.

Beyond that, I'm happy to help with testing, reviews, or anything else that's useful.

**Comment by bringlein** [comment]
CC @bohnstingl (he did look a lot into the tensor layouts for the vLLM plugin)

**Comment by bohnstingl** [comment]
I think the proper place for such weight transformations is the `process_weights_after_loading` function. For example, this is currently done for the [SpyreParallelLMHead](https://github.com/torch-spyre/spyre-inference/blob/6287b104013a0ca93b6361cbe9e4b7a0676e3696/spyre_inference/custom_ops/parallel_lm_head.py#L44).

This being said, we currently mostly rely on the parameter layout from upstream vLLM and it might be worthwhile checking whether a different layout would yield performance improvements for spyre-inference. I created an [issue](https://github.com/torch-spyre/spyre-inference/issues/469) for this.

**Comment by dmholtz** [comment]
As promised, this is the summary of affected layers in Granite with vllm (spyre-inference): https://github.com/torch-spyre/spyre-inference/issues/470
